torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1556,8 +1557,18 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::collective(
       //
       // See [Sync Streams].
       if (!avoidRecordStreams_) {
-        c10::cuda::CUDACachingAllocator::recordStream(
-            inputs[i].storage().data_ptr(), ncclStream);
+        // TODO: fix for sparse tensors
+        // C++ exception with description "Cannot access storage of
+        // SparseTensorImpl
+        if (!inputs[i].is_sparse()) {
+          c10::cuda::CUDACachingAllocator::recordStream(
+              inputs[i].storage().data_ptr(), ncclStream);
+        } else {
+          c10::cuda::CUDACachingAllocator::recordStream(
+              inputs[i].values().storage().data_ptr(), ncclStream);
+          c10::cuda::CUDACachingAllocator::recordStream(
```

Comment:
nit: I think the lack of this was a bug that stumped us for a while, so can you add a comment why this is needed? 

IIRC - since the tensor will be used on a different (the nccl) stream, we need to call recordStream to ensure the tensor is not prematurely freed?